receiver/carbonreceiver/protocol/path_parser_helper.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package protocol
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+)
+
+// PathParser implements the code needed to handle only the <metric_path> part of
+// a Carbon metric line:
+//
+// 	<metric_path> <metric_value> <metric_timestamp>
+//
+// See https://graphite.readthedocs.io/en/latest/feeding-carbon.html#the-plaintext-protocol,
+// for more information.
+//
+// The type PathParserHelper implements the common code for parsers that differ
+// only by the way that they handle the <metric_path>.
+type PathParser interface {
+	// ParsePath converts the <metric_path> of a Carbon line (see Parse function
+	// for description of the full line). The return values are:
+	//
+	// - name: the metric name extracted/generated by the parser.
+	// - keys: the label keys extracted/generated by the parser.
+	// - values: the label values extracted/generated by the parser.
+	// - forceCumulative: forces the metric to be a cumulative counter instead of
+	//   a gauge.
```

Comment:
With so many return values is it worth defining a struct for this and returning the struct by value?

Reply:
There was a noticeable (~10%) penalty because of that, but I liked the idea of having a struct around those but the only way to avoid the perf penalty was to pass the struct by reference to the function.

Reply:
btw, passing the struct by reference gets ~10% perf gain over the original code.

Reply:
> btw, passing the struct by reference gets ~10% perf gain over the original code.

Can you please confirm that this is true for long-running case when memorry allocations and GC is in sustained mode? I wonder if we may see perf gain when tests are short just because GC does not kick in at all.

Reply:
I ran the tests for 10s and never mixed the versions in the same run

Reply:
Sounds good, thanks for confirming.